app_allocator/classes/needs_queue.py

```diff
@@ -0,0 +1,78 @@
+from copy import deepcopy
+from math import log
+from app_allocator.classes.assignments import can_be_assigned
+
+
+class NeedsQueue(object):
+    def __init__(self, needs={}, count=1):
+        self.needs = deepcopy(needs)
+        self.count = count
+        self.counts = {}
+        self.items = []
+        self.name = str(self)
+        self.assignments = {}
+
+    def __str__(self):
+        return "Q(%s)[%s]" % (",".join([str(need) for need in self.needs]),
+                              len(self.items))
+
+    def remaining(self):
+        return len(self.items)
+
+    def add_application(self, application, at_front=True):
+        if at_front:
+            self.items.insert(0, application)
+        else:
+            self.items.append(application)
+
+    def move_to_end(self, application):
+        if application in self.items:
+            self.items.remove(application)
+            self.items.append(application)
+
+    def remove_application(self, application):
+        self.items.remove(application)
+        if application in self.assignments:
+            self.assignments.pop(application)
+
+    def remove_assignment(self, application, judge):
+        assignments = self.assignments.get(application, [])
+        if judge in assignments:
+            assignments.remove(judge)
+
+    def work_left(self):
+        return len(self.items)
+
+    def judge_value(self, judge):
```

Comment:
This is a little bit peculiar. It seems wrong to me that the value of this queue depends on the state of the first application in that queue, when there's no particular effort to ensure that the first app in the queue is the neediest. 
One possible error state that this could produce would be if the first app in the queue ends up with a low value for `value`, and there are needier applications behind it. Those applications would then be blocked from getting reads until that relatively needless app got selected. 
How do we guard against that condition?

Not a blocker, since this is not production code, but I wanted to make sure it was noted.

Reply:
It's not based on the first app in the queue.  It's based on the application this judge would get if we take an application from this queue.  That seems like exactly the right thing to be evaluating.  If their are needier apps behind the one that was chosen, then that's an issue with `_next_application`.  Currently this heuristic does just look for the first application that meets the z-score criteria and that the judge has not judged.  If the application at the front of the queue has no assignments in flight, then the first one is fine.  The tricky part is when the front element has been assigned.  It then becomes a question of how hard we want to look for a better match.  If we look exhaustively, then there's significantly less benefit to having the queues.  One option would be to figure out for each queue what the greatest possible value is (which would be for an unassigned application) and not search queues whose greatest possible value is less than the value of the current candidate.  There would still be cases where you would have to do an exhaustive search to find the "best".  I think solving this question is outside the scope of this PR.

Reply:
I agree that the reassignment condition is where the potential problem arises, not that concerned about whether the "real" problem is here or in `_next_application` - it arises most clearly here, this is where we'd see the bug if it arises, so this is where I mentioned it. 

I agree that the solution is not required for this PR, hence not a blocker, but it seems like a solvable problem that is worth solving.